emails/views.py

```diff
@@ -1784,84 +1738,331 @@ def _handle_complaint(message_json: AWS_SNSMessageJSON) -> HttpResponse:
+            "_get_complaint_data: Unexpected message format",
+            extra={"missing_key": key, "found_keys": ",".join(sorted(source.keys()))},
+        )
+        return data_type(), False
+
+    raw_recipients, has_cr = get_or_log("complainedRecipients", complaint, list)
+    complained_recipients = []
+    no_entries = True
+    for entry in raw_recipients:
+        no_entries = False
+        raw_email_address, has_email = get_or_log("emailAddress", entry, str)
+        if has_email:
+            email_address = parseaddr(raw_email_address)[1]
+            extra = {
+                key: value for key, value in entry.items() if key != "emailAddress"
+            }
+            complained_recipients.append((email_address, extra))
+    if has_cr and no_entries:
+        logger.error("_get_complaint_data: Empty complainedRecipients")
+
+    mail, has_mail = get_or_log("mail", message_json, dict)
+    if has_mail:
+        commonHeaders, has_ch = get_or_log("commonHeaders", mail, dict)
+    else:
+        commonHeaders, has_ch = {}, False
+    if has_ch:
+        raw_from_addresses, _ = get_or_log("from", commonHeaders, list)
+    else:
+        raw_from_addresses = []
+    from_addresses = [parseaddr(addr)[1] for addr in raw_from_addresses]
+
+    feedback_type, _ = get_or_log("complaintFeedbackType", complaint, str)
+
+    # Only present when destination is on account suppression list
+    subtype = complaint.get("complaintSubType", "")
+    # Only present for feedback reports
+    user_agent = complaint.get("userAgent", "")
+
+    return RawComplaintData(
+        complained_recipients, from_addresses, subtype, user_agent, feedback_type
+    )
+
+
+class Complainer(TypedDict):
+    user: User
+    found_in: Literal["complained_recipients", "from_header", "all"]
+    domain: str
+    extra: dict[str, Any] | None
+    masks: list[RelayAddress | DomainAddress]
+
+
+def _gather_complainers(
+    complaint_data: RawComplaintData,
+) -> tuple[list[Complainer], int]:
+    """
+    Fetch Relay Users and masks from the complaint data.
+
+    This matches data from an AWS SES Complaint Notification (as extracted by
+    _get_complaint_data()) to the Relay database, and returns the Users,
+    RelayAddresses, and DomainAddresses, as well as status and extra data.
+
+    If the complaint came from the AWS SES complaint simulator, detect
+    developer_mode and move forward with the developer's User data.
+    """
+
+    users: dict[int, Complainer] = {}
+    unknown_complainer_count = 0
+    for email_address, extra_data in complaint_data.complained_recipients:
+        local, domain = email_address.split("@", 1)
+
+        # If the complainer is the AWS SES complaint simulation, assume that
+        # it was send by a user with the developer_mode flag. Look for
+        # a mask that matches the embedded mask metrics_id, and use
+        # the related user's email instead of the AWS simulator address.
+        # See docs/developer_mode.md
```

Comment:
suggestion (blocking): Can we rename the doc file to `docs/developer_mode.md` like this comment says? That will help future ENGRs (including us) find the doc file when searching the repo for the name of this flag.

Reply:
Yes, nice catch!